### PR TITLE
[PyTorch] Experimental FP8 tensor class

### PR DESCRIPTION
**Pull request by timmoon10**
This FP8 tensor class is based on the implementation at https://github.com/facebookexperimental/protoquant/tree/fp8_poc and is primarily oriented toward enabling efficient FP8 support in Apex's `DistributedFusedAdam`. See https://github.com/NVIDIA/NeMo/pull/7469 and https://github.com/NVIDIA/NeMo/pull/7565.

CC @sudhakarsingh27 @ksivaman

### CODE REVIEW

File: transformer_engine/pytorch/fp8.py [hunk 1]
```diff
@@ -515,7 +515,7 @@ def _update_amax_history(amax_history: torch.Tensor) -> torch.Tensor:
     return amax_history
 
 
-@jit_fuser
+# @jit_fuser
```
Comment (ptrendx) [inline transformer_engine/pytorch/fp8.py:518]: @sudhakarsingh27 why did you remove the fusion of these functions?

Comment (timmoon10) [reply transformer_engine/pytorch/fp8.py:518]: We can probably restore them. I suspect it's a debugging artifact.

File: transformer_engine/pytorch/float8_tensor.py [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+    The tensor presents as having a standard, higher-precision dtype,
+    but the data itself is (scaled) FP8. For most tensor operations,
+    the data will be cast to the nominal dtype before performing the
+    operation.
+
+    Changes to the FP8 scaling factors, e.g. from the FP8 recipe, are
+    handled outside this class. If a tensor is initialized with an FP8
+    metadata object, it extracts the information it needs so it isn't
+    affected by later changes in the FP8 metadata (although its design
+    does cause us to leak some subtle side-effects into FP8 metadata).
```
Comment (ptrendx) [inline transformer_engine/pytorch/float8_tensor.py:207]: This doc is not really correct since we are holding a view to the meta, right?

File: transformer_engine/pytorch/float8_tensor.py [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+                self._fp8_meta["recipe"],
+                fprop_tensor=self._fp8_meta_forward,
+            )
+        if self._fp8_dtype is None:
+            raise ValueError(
+                "Attempted to initialize Float8Tensor without specifying FP8 dtype"
+            )
+
+        # Cached transpose
+        self._transpose: Optional[Float8Tensor] = None
```
Comment (ptrendx) [inline transformer_engine/pytorch/float8_tensor.py:295]: Caching of the transpose should be configurable somehow.

File: transformer_engine/pytorch/float8_tensor.py [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+        return self.from_float8(dtype=torch.float32)
+
+    def bfloat16(self) -> torch.Tensor:
+        return self.from_float8(dtype=torch.bfloat16)
+
+    def half(self) -> torch.Tensor:
+        return self.from_float8(dtype=torch.float16)
+
+    def cpu(self) -> torch.Tensor:
+        return self.from_float8().cpu()
```
Comment (ptrendx) [inline transformer_engine/pytorch/float8_tensor.py:402]: Shouldn't it use `self.dtype` for this cast?

File: transformer_engine/pytorch/float8_tensor.py [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+        if other is self:
+            # Note: expand_as is hackily used to create dummy autograd nodes
+            # and access the backward graph (see
+            # https://github.com/pytorch/pytorch/blob/238fb660851268f44ff88127887041fea352fe48/torch/nn/parallel/distributed.py#L1026).
+            # We equally hackily add a dummy function to handle this
+            # case.
+            return _IdentityFunc.apply(self)
+        return super().expand_as(other)
+
+    def transpose(self, dim0: int = 0, dim1: int = 1) -> Float8Tensor:
```
Comment (ptrendx) [inline transformer_engine/pytorch/float8_tensor.py:417]: Maybe add an optional argument here that says whether to cache the result?

File: transformer_engine/pytorch/fp8.py [hunk 6]
```diff
@@ -454,6 +460,41 @@ def restore_fp8_meta_tensors(fp8_meta: Dict[str, Any]) -> None:
         fp8_meta["scaling_fwd"].scale_inv = fp8_meta["updated_scale_inv_fwd"]
 
 
+@contextmanager
+def fp8_init(enabled: bool = False) -> None:
+    """
+    Context manager for FP8 initialization of parameters.
+
```
Comment (ptrendx) [inline transformer_engine/pytorch/fp8.py:467]:
```suggestion
Example usage:

```

File: transformer_engine/pytorch/module/base.py [hunk 7]
```diff
@@ -777,23 +791,42 @@ def get_fp8_weights_empty_tensors(
...
+    def state_dict(self, *args, **kwargs) -> Dict:
+        """Get dictionary containing module state"""
+        state = super().state_dict(*args, **kwargs)
+
+        # Convert Float8Tensors to plain tensors
+        # Note: Float8Tensors don't serialize well, especially if they
+        # contain references to FP8 metadata.
+        for key, val in state.items():
+            if isinstance(val, Float8Tensor):
+                state[key] = val.from_float8()
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/base.py:827]: That will result in FP32 tensor being stored, right? This should use the dtype of the FP8 tensor, maybe `from_float8` should actually by default use `self.dtype`?

File: transformer_engine/pytorch/module/layernorm_linear.py [hunk 8]
```diff
@@ -79,10 +80,11 @@ def forward(
         fwd_ln_sm_margin: int,
         bwd_ln_sm_margin: int,
         zero_centered_gamma: bool,
+        normalization: str,
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/layernorm_linear.py:83]: Random question: why change the spot of the normalization string?

File: transformer_engine/pytorch/float8_tensor.py (hunk 4 above)
Comment (timmoon10) [reply transformer_engine/pytorch/float8_tensor.py:402]: The default behavior of `from_float8` is to use the tensor's nominal dtype.

File: transformer_engine/pytorch/float8_tensor.py (hunk 5 above)
Comment (timmoon10) [reply transformer_engine/pytorch/float8_tensor.py:417]: Alternatively, we could do something like:
```python
def transpose(self):
    if self._transpose is not None:
        return self._transpose
    return super().transpose()
```
The cached transpose should be allocated and initialized externally, e.g. in the TE modules.

File: transformer_engine/pytorch/fp8.py [hunk 9]
```diff
@@ -477,6 +483,41 @@ def restore_fp8_meta_tensors(fp8_meta: Dict[str, Any]) -> None:
         fp8_meta["scaling_fwd"].scale_inv = fp8_meta["updated_scale_inv_fwd"]
 
 
+@contextmanager
+def fp8_init(enabled: bool = False) -> None:
```
[the remark this answers is not in the document]
Comment (sudhakarsingh27) [reply transformer_engine/pytorch/fp8.py:487]: Could we rename it to something like `fp8_autocast_init`, to differentiate from the other `fp8_init` that's in `base.py` file? Also, the other `fp8_init` could be more descriptive like `fp8_metadata_init` or similar.

File: transformer_engine/pytorch/fp8.py [hunk 10]
```diff
@@ -477,6 +483,41 @@ def restore_fp8_meta_tensors(fp8_meta: Dict[str, Any]) -> None:
...
+                      change
+
+                  This functionality is *EXPERIMENTAL*.
+    """
+    try:
+        _fp8_parameters = FP8GlobalStateManager.FP8_PARAMETERS
+        FP8GlobalStateManager.FP8_PARAMETERS = enabled
+        yield
+    finally:
+        FP8GlobalStateManager.FP8_PARAMETERS = _fp8_parameters # pylint: disable=used-before-assignment
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/fp8.py:518]: This could potentially revert back to "non-fp8" state upon CM exit. 
So, doing something like:
```
with fp8_init(enabled=True):
            model = transformer_engine.pytorch.Linear(768, 768) # model init in true FP8

with fp8_autocast(enabled=True):
            output = model(input) # model doesn't run in true FP8 <-----
```
because, let's say, `Linear` would call `self.primary_weights_in_fp8 = FP8GlobalStateManager.is_fp8_parameters()` which might return false once we're out of the `fp8_init` context manager? 

Unless the default value of `FP8GlobalStateManager.FP8_PARAMETERS` is `True` which wouldn't make sense either

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/fp8.py:518]: Or is the intended use something like:
```
with fp8_init(enabled=True):
            model = transformer_engine.pytorch.Linear(768, 768) # model init in true FP8

            with fp8_autocast(enabled=True):
                            output = model(input) # model also runs in true FP8 
```
Which is fine too, but a bit inflexible? Also, will need to document this usage.

File: transformer_engine/pytorch/fp8.py (hunk 9 above)
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/fp8.py:487]: Could we rename it to something like `fp8_model_init`, to differentiate from the other `fp8_init` that's in `base.py` file? Also, the other `fp8_init` could be more descriptive like `fp8_metadata_init` or similar.

File: transformer_engine/pytorch/fp8.py (hunk 10 above)
Comment (ptrendx) [reply transformer_engine/pytorch/fp8.py:520]: Not sure I understand. `self.primary_weights_in_fp8` is set during constructor which happens inside `fp8_init`. Afterwards it is used irrespective of the FP8_PARAMETERS value of the context manager...

Comment (ksivaman) [reply transformer_engine/pytorch/fp8.py:520]: That's right. Since we are only using the `FP8GlobalStateManager.has_fp8_parameters()` query during initialization, the value of `self.primary_weights_in_fp8` can never change during the forward execution. The only way to toggle fp8 behavior would be using the `fp8_autocast` API, which is the intended method of use.

Comment (ksivaman) [reply transformer_engine/pytorch/fp8.py:520]: @sudhakarsingh27 Your intended snippet would work but the canonical (and recommended) way of going about it would be:
```
with fp8_init(enabled=True):
        model = transformer_engine.pytorch.Linear(768, 768) # model init in true FP8

with fp8_autocast(enabled=True):
        output = model(input) # model also runs in true FP8
```

Comment (ksivaman) [reply transformer_engine/pytorch/fp8.py:520]: We have guards to ensure that when model is initialized in the `fp8_init` context, the forward execution happens with `fp8_autocast(enabled=True)`.

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/fp8.py:520]: Ah you're right! I completely misunderstood `fp8_init` usage

File: transformer_engine/pytorch/module/linear.py [hunk 11]
```diff
@@ -609,18 +634,31 @@ def __init__(
...
             stride=1,
         )
 
+        if self.primary_weights_in_fp8:
+            self.fp8_init()
+            self.fp8_meta["update_amax_and_scale_fwd"] = True
+
+            self.weight_tensor = Float8Tensor.to_float8(
+                temp_weight,
+                fp8_meta=self.fp8_meta,
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/module/linear.py:656]: With `fp8_init`, there's no option to specify the "recipe" but the weight tensor would store the reference to `fp8_meta` created during module initialization, which I guess will always point to the default recipe config. Right, @ksivaman? (@timmoon10 for reference)

File: transformer_engine/pytorch/float8_tensor.py [hunk 12]
```diff
@@ -457,21 +457,21 @@ def transpose(
...
-        # Use optimized kernel for basic 2D transpose
-        # TODO Support differentiation # pylint: disable=fixme
-        return Float8Tensor.make_like(
-            self,
-            data=tex.fp8_transpose(
-                self._data.contiguous().detach(),
-                self._fp8_dtype,
-            ),
-        )
+        if cache or self._transpose is None:
```
Comment (timmoon10) [inline transformer_engine/pytorch/float8_tensor.py:472]: `cache == True` at this point:
```suggestion
        if self._transpose is None:
```

File: transformer_engine/pytorch/float8_tensor.py [hunk 13]
```diff
@@ -0,0 +1,654 @@
...
+
+            # Cast to FP8
+            tex.cast_to_fp8_noalloc(
+                src.view(1,-1),
+                dst._scale_inv.reciprocal(),
+                dst._data.view(1,-1),
+                torch.empty_like(dst._scale_inv),  # amax
+                dst._scale_inv,
+                dst._fp8_dtype,
+            )
```
Comment (timmoon10) [inline transformer_engine/pytorch/float8_tensor.py:542]: I've noticed we never update `scale_inv`, even if fp8_meta changes. I think updating within `copy_` should handle this:
```suggestion
            # Update scaling factor if FP8 meta tensors are available
            if dst._fp8_meta is None:
                scale = dst._scale_inv.reciprocal()
            else:
                fp8_meta_key = FP8GlobalStateManager.get_meta_tensor_key(
                    forward=dst._fp8_meta_forward,
                )
                scale = dst._fp8_meta[fp8_meta_key].scale[dst._fp8_meta_index]
                dst._scale_inv = scale.detach().view(1).reciprocal()

            # Cast to FP8
            tex.cast_to_fp8_noalloc(
                src.view(1,-1),
                scale,
                dst._data.view(1,-1),
                torch.empty_like(dst._scale_inv),  # amax
                dst._scale_inv,
                dst._fp8_dtype,
            )
```

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/float8_tensor.py:556]: Ah right, there used to be a swapping logic for `scale_inv` when we first discussed this. 
Having it in `copy_` seems fine but I guess we might need it in other in-place operations like `add_`? So, we could have updating `scale_inv` as part of some "inplace_op_hook". Wdyt?

Comment (timmoon10) [reply transformer_engine/pytorch/float8_tensor.py:556]: The other in-place ops call `copy_` when casting the result back to FP8, so shouldn't `copy_` be sufficient?

Comment (ksivaman) [reply transformer_engine/pytorch/float8_tensor.py:556]: Yes `copy_` should be sufficient since it's the end point of inplace ops. Trying the approach of storing `_scale_inv` in a mutable data structure as opposed to a direct attribute in `Float8Tensor`.

Comment (sudhakarsingh27) [reply transformer_engine/pytorch/float8_tensor.py:556]: Okay `copy_` seems to be the only realized in-place op in Float8Tensor class, so this sounds good then!

**Review by ptrendx — APPROVED** [review]
Approving as experimental. We will iterate upon this in the next release.

File: transformer_engine/pytorch/fp8.py [hunk 14]
```diff
@@ -477,9 +483,45 @@ def restore_fp8_meta_tensors(fp8_meta: Dict[str, Any]) -> None:
...
+    ----------
+    enabled: bool, default = `True`
+             when enabled, Transformer Engine modules created inside this `fp8_model_init`
+             region will hold only FP8 copies of its parameters, as opposed to the default
+             behavior where both higher precision and FP8 copies are present. Setting this
+             option to `True` may result in lower memory consumption and is especially
+             useful for scenarios like:
+
+             * full model training using optimizer with master weights, where the high
+               precision copies of weights are already present in the optimizer.
```
Comment (carmocca) [inline transformer_engine/pytorch/fp8.py:508]: How does this look in practice? If the model will be initialized directly with fp8 weights, how does the optimizer get high-precision copies?
